File: go/pkg/protocol/metadata.go
package protocol

type SchemaInfo struct {
	Name string `json:"name" msgpack:"name"`
}

type TableType string

const (
	TableTypeTable            TableType = "table"
	TableTypeView             TableType = "view"
	TableTypeMaterializedView TableType = "materialized_view"
)

type TableInfo struct {
	Name     string    `json:"name" msgpack:"name"`
	Schema   string    `json:"schema" msgpack:"schema"`
	Type     TableType `json:"type" msgpack:"type"`
	RowCount *int64    `json:"row_count,omitempty" msgpack:"row_count,omitempty"`
	Comment  string    `json:"comment,omitempty" msgpack:"comment,omitempty"`
}

type ColumnInfo struct {
	Name            string      `json:"name" msgpack:"name"`
	DataType        string      `json:"data_type" msgpack:"data_type"`
	GenericType     GenericType `json:"generic_type,omitempty" msgpack:"generic_type,omitempty"`
	Nullable        bool        `json:"nullable" msgpack:"nullable"`
	DefaultValue    *string     `json:"default_value,omitempty" msgpack:"default_value,omitempty"`
	MaxLength       *int32      `json:"max_length,omitempty" msgpack:"max_length,omitempty"`
	Precision       *int32      `json:"precision,omitempty" msgpack:"precision,omitempty"`
	Scale           *int32      `json:"scale,omitempty" msgpack:"scale,omitempty"`
	IsPrimaryKey    bool        `json:"is_primary_key" msgpack:"is_primary_key"`
	IsAutoIncrement bool        `json:"is_auto_increment" msgpack:"is_auto_increment"`
	Comment         string      `json:"comment,omitempty" msgpack:"comment,omitempty"`
	OrdinalPosition int32       `json:"ordinal_position" msgpack:"ordinal_position"`
}

type IndexType string

const (
	IndexTypeBTree    IndexType = "btree"
	IndexTypeHash     IndexType = "hash"
	IndexTypeGin      IndexType = "gin"
	IndexTypeGist     IndexType = "gist"
	IndexTypeFulltext IndexType = "fulltext"
	IndexTypeSpatial  IndexType = "spatial"
	IndexTypeOther    IndexType = "other"
)

type IndexInfo struct {
	Name    string        `json:"name" msgpack:"name"`
	Unique  bool          `json:"unique" msgpack:"unique"`
	Primary bool          `json:"primary" msgpack:"primary"`
	Type    IndexType     `json:"type,omitempty" msgpack:"type,omitempty"`
	Columns []IndexColumn `json:"columns" msgpack:"columns"`
	Comment string        `json:"comment,omitempty" msgpack:"comment,omitempty"`
}

type IndexColumn struct {
	Name         string    `json:"name" msgpack:"name"`
	Position     int32     `json:"position" msgpack:"position"`
	Order        SortOrder `json:"order,omitempty" msgpack:"order,omitempty"`
	PrefixLength *int32    `json:"prefix_length,omitempty" msgpack:"prefix_length,omitempty"`
}

type ForeignKeyAction string

const (
	FKActionCascade    ForeignKeyAction = "cascade"
	FKActionSetNull    ForeignKeyAction = "set_null"
	FKActionSetDefault ForeignKeyAction = "set_default"
	FKActionRestrict   ForeignKeyAction = "restrict"
	FKActionNoAction   ForeignKeyAction = "no_action"
)

type ForeignKeyInfo struct {
	Name              string           `json:"name" msgpack:"name"`
	Columns           []string         `json:"columns" msgpack:"columns"`
	ReferencedSchema  string           `json:"referenced_schema" msgpack:"referenced_schema"`
	ReferencedTable   string           `json:"referenced_table" msgpack:"referenced_table"`
	ReferencedColumns []string         `json:"referenced_columns" msgpack:"referenced_columns"`
	OnUpdate          ForeignKeyAction `json:"on_update,omitempty" msgpack:"on_update,omitempty"`
	OnDelete          ForeignKeyAction `json:"on_delete,omitempty" msgpack:"on_delete,omitempty"`
}

File: rust/crates/protocol/src/metadata.rs
use serde::{Deserialize, Serialize};

use crate::common::{GenericType, SortOrder};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SchemaInfo {
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TableInfo {
    pub name: String,
    pub schema: String,
    #[serde(rename = "type")]
    pub table_type: TableType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub row_count: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comment: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TableType {
    Table,
    View,
    MaterializedView,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ColumnInfo {
    pub name: String,
    pub data_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generic_type: Option<GenericType>,
    pub nullable: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_value: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_length: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub precision: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scale: Option<i32>,
    #[serde(default)]
    pub is_primary_key: bool,
    #[serde(default)]
    pub is_auto_increment: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comment: Option<String>,
    pub ordinal_position: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IndexInfo {
    pub name: String,
    pub unique: bool,
    #[serde(default)]
    pub primary: bool,
    #[serde(rename = "type")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub index_type: Option<IndexType>,
    pub columns: Vec<IndexColumn>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comment: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum IndexType {
    Btree,
    Hash,
    Gin,
    Gist,
    Fulltext,
    Spatial,
    Other,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IndexColumn {
    pub name: String,
    pub position: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub order: Option<SortOrder>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prefix_length: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ForeignKeyInfo {
    pub name: String,
    pub columns: Vec<String>,
    pub referenced_schema: String,
    pub referenced_table: String,
    pub referenced_columns: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub on_update: Option<ForeignKeyAction>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub on_delete: Option<ForeignKeyAction>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ForeignKeyAction {
    Cascade,
    SetNull,
    SetDefault,
    Restrict,
    NoAction,
}
